### PR TITLE
Support glEGLImageTargetTexture2DOES for Android GLES

### PR DESCRIPTION
**Pull request by lfedorov**
After RenderDoc began to strictly check extensions, android developers were faced with the fact that due to the unsupported glEGLImageTargetTexture2DOES extension, they lost the ability to use the RenderDoc. Most modern android games use this extension to display videos or ads.

Here is support of this extension for Android platform.

Functionality can be tested in this sample: http://l1f.ru/contributing/renderdoc/external_texture/external_texture_debug.apk 

Also in the serialiser.cpp there is a small fix of chunks processing when loading from xml+zip.

### CODE REVIEW

File: CMakeLists.txt
Comment (Squareys) [inline CMakeLists.txt]: The comment here will probably need updating.
Why is the upgrade necessary? (It could narrow Android OS version compatibility, possibly, if I understand correctly.)

File: renderdoc/api/replay/replay_enums.h
Comment (Squareys) [inline renderdoc/api/replay/replay_enums.h]:
```suggestion
  An external texture.
```

File: renderdoc/driver/gl/CMakeLists.txt
Comment (Squareys) [inline renderdoc/driver/gl/CMakeLists.txt]: Be careful to use spaces indentation to match the rest of the file😌

File: renderdoc/driver/gl/gl_dispatch_table.h
Comment (Squareys) [inline renderdoc/driver/gl/gl_dispatch_table.h]: This newline seems superfluous/unrelated

File: renderdoc/driver/gl/gl_initstate.cpp
Comment (Squareys) [inline renderdoc/driver/gl/gl_initstate.cpp]:
```suggestion
        // update width from here as it's authoritative - the texture might have been resized in
        // multiple rebinds that we will not have serialised before.
```

File: renderdoc/driver/gl/wrappers/gl_external_texture.cpp
Comment (Squareys) [inline renderdoc/driver/gl/wrappers/gl_external_texture.cpp:25]: The MIT headers are missing here :)

File: renderdoc/driver/gl/wrappers/gl_external_texture.cpp
Comment (Squareys) [inline renderdoc/driver/gl/wrappers/gl_external_texture.cpp]: This empty comment line can probably be either removed or turned into a newline.

File: renderdoc/driver/gl/wrappers/gl_external_texture.cpp
Comment (Squareys) [inline renderdoc/driver/gl/wrappers/gl_external_texture.cpp]: Renderdoc code seems to prefer to put comments on the line above rather than appending at the end of the line.

In case you haven't done so yet, it seems that clang format can take care of most of the coding style:
https://github.com/baldurk/renderdoc/blob/v1.x/docs/CONTRIBUTING/Developing-Change.md

**Review by Squareys — CHANGES_REQUESTED** [review]
Hey @lfedorov!

Thank you so much for working on this! 🎉 I went through the changes on a superficial level (code formatting / plausibility / spelling). I don't think I'm qualified to comment on the implementation itself.

(Hope my comments are somewhat helpful, Baldur, and I'm not getting in the way. Let me know otherwise.)

Best,
Jonathan

File: CMakeLists.txt
Comment (lfedorov) [reply CMakeLists.txt]: Yep. I will fix this comment.
Native Buffer is available only from API level 26, so it need to be updated or need to replace with something else.
https://developer.android.com/ndk/reference/group/a-hardware-buffer

**Comment by baldurk** [comment]
@Squareys do not leave reviews on pull requests from repositories that you are not authorised to. I'm amazed that github even allows this, apparently this is enabled by default and I have now changed that.

It is rude and arrogant to write a review on a repository you have absolutely no relation to nor had any interaction with, leave that to the repository maintainers or anyone they ask to review the code otherwise you will only cause confusion - people submitting PRs do not know if you are one of the maintainers or a complete stranger.

File: renderdoc/api/replay/replay_enums.h
Comment (lfedorov) [reply renderdoc/api/replay/replay_enums.h]: fixed

File: renderdoc/driver/gl/CMakeLists.txt
Comment (lfedorov) [reply renderdoc/driver/gl/CMakeLists.txt]: fixed.. it seems clang-format did not process it

File: renderdoc/driver/gl/gl_dispatch_table.h
Comment (lfedorov) [reply renderdoc/driver/gl/gl_dispatch_table.h]: fixed

File: renderdoc/driver/gl/gl_initstate.cpp
Comment (lfedorov) [reply renderdoc/driver/gl/gl_initstate.cpp]: fixed

File: renderdoc/driver/gl/wrappers/gl_external_texture.cpp
Comment (lfedorov) [reply renderdoc/driver/gl/wrappers/gl_external_texture.cpp]: fixed

File: renderdoc/driver/gl/wrappers/gl_external_texture.cpp
Comment (lfedorov) [reply renderdoc/driver/gl/wrappers/gl_external_texture.cpp]: fixed.. and I also fix some nullptr and auto's in my code. Thank you!

File: renderdoc/driver/gl/wrappers/gl_external_texture.cpp
Comment (lfedorov) [reply renderdoc/driver/gl/wrappers/gl_external_texture.cpp:25]: fixed

**Comment by lfedorov** [comment]
@baldurk Sorry for that, but @Squareys was very helpful in finding corner cases with this implementations, and these suggestions were quite useful. I've already fix them.

@baldurk Sure, I also wait for your review, it seems there is something to improve here.

File: CMakeLists.txt
Comment (baldurk) [inline CMakeLists.txt:66]: I'm not sure what this is intended to do since it only changes the default platform if none is specified, but this does not stop anyone from continuing to explicitly specify android-21. This doesn't change RenderDoc's minspec, all this does is means that CI and anyone building by default will no longer target the minspec correctly and it may hide problems.

If you want to build targeting a higher version you can, you can specify the `ANDROID_PLATFORM` parameter to cmake.

File: renderdoc/api/replay/replay_enums.h
Comment (baldurk) [inline renderdoc/api/replay/replay_enums.h:739]: External textures are not a type of texture, so there should not be a new enum in this list. External textures are normal textures (so one of the other types in this list) but with a different creation method that refers elsewhere outside the API. All APIs have external texture support in RenderDoc, including multiple existing extensions for importing external textures in GL, but they do not define a new texture type.

File: renderdoc/data/glsl/glsl_globals.h
Comment (baldurk) [inline renderdoc/data/glsl/glsl_globals.h:170]: Adding a new resource type here doesn't make sense without at least corresponding shader changes. Otherwise setting a new texture slot in code will refer to a non-existing texture binding and cause textures to not display correctly. As above when talking about `TextureType`, external textures should not be treated as a different type of texture and instead should be accessed via `RESTYPE_TEX1D`, `RESTYPE_TEX2D`, `RESTYPE_TEX2DMS` etc depending on their actual type.

File: renderdoc/driver/gl/wrappers/gl_external_texture.cpp
Comment (baldurk) [inline renderdoc/driver/gl/wrappers/gl_external_texture.cpp:98]: The replay should not be based around creating actual external textures. That's outside the scope of what RenderDoc does and exposes a potential unknown API hole with creating a compatible external texture.

If you look at `gl_interop_funcs.cpp` for how the other external sharing GL APIs are implemented, you'll see that e.g. in `WrappedOpenGL::Serialise_wglDXRegisterObjectNV` for windows or `WrappedOpenGL::Serialise_glTextureStorageMem1DEXT` which is platform agnostic, the textures are  created just as plain textures but with properties to mimic the original imported texture.

The replay of this extension should function the same, not depending on any native APIs like this but during capture snapshotting their properties and on replay creating a "fake" standin for the external texture. This then means RenderDoc only needs to cover the API that is in scope (GL ES in this case) and means it behaves more predictably and understandably.

File: renderdoc/serialise/serialiser.cpp
Comment (baldurk) [inline renderdoc/serialise/serialiser.cpp:590]: Please do not make unrelated changes to core serialisation code.

**Review by baldurk — CHANGES_REQUESTED** [review]
Thanks for this PR. I've not done a detailed review yet as I looked at it at a high level and there are some architectural refactors that I'd like you to make, so there's no reason to get into fine details until they were done.

I've put more information in the comments above, but the main two things are to not special-case external textures beyond what's internally necessary for implementation details, and to not actually recreate textures as external on replay but instead create normal textures as stand-ins to avoid external dependencies.

**Comment by baldurk** [comment]
btw as a minor point of clarification, glEGLImageTargetTexture2DOES is not an extension it's an extension _function_, the extension itself is OES_EGL_image or OES_EGL_image_external. This is important because not all extensions have only one function, and within your changes you have not added either extension to the list of allowed extensions.

Implementing this function alone will do nothing for a correctly written program since it will still see the extension reported as unavailable. If you are seeing a difference in behaviour with this PR that means the program you're testing with is broken and not functioning correctly.

File: renderdoc/serialise/serialiser.cpp
Comment (lfedorov) [reply renderdoc/serialise/serialiser.cpp:590]: Do you think I need to make another PR for that one?

**Comment by lfedorov** [comment]
@baldurk I got your point about the texture type. I'll try to make these updates.
About implementing ony one function from extenstion: As I understand it will works correctly if application does not check ext availability with RenderDoc. It's on developer side. Will you approve PR if it implement only glEGLImageTargetTexture2DOES in this case (if other stuff is OK)?

File: CMakeLists.txt
Comment (lfedorov) [reply CMakeLists.txt:66]: Do we have place to specify android minspec for that? But anyway, I think I don't need it if I rewrite the code with your suggestions...

File: renderdoc/serialise/serialiser.cpp
Comment (baldurk) [reply renderdoc/serialise/serialiser.cpp:590]: Yes please keep each PR related to a single set of functionality, if you have another change you want to propose please make it separately after this one is resolved.

**Comment by baldurk** [comment]
RenderDoc does not intend to handle broken applications and we don't write code assuming applications will misbehave and then work by coincidence. Any application that tries to use functionality without checking if the extension is supported is not correct.

You must implement the entire extension that you are intending to support and report it, to ensure that _valid_ applications will work correctly. As long as everything in the extension is working that shouldn't be an issue, and I only mentioned it to make sure that you remembered to allow the extension itself as there are no other functions to implement. If you have not looked at the entire extension to make sure you are implementing it correctly you should do that, as it will be reported as supported and the entirety of it is expected to work.

File: CMakeLists.txt
Comment (baldurk) [reply CMakeLists.txt:66]: Currently the min spec is Android 6 and only mentioned I think on the [documentation page](https://renderdoc.org/docs/how/how_android_capture.html) since the minspec is checked at runtime and shown when trying to connect to the device.

I don't know why you were changing this default so I can't answer if it's still needed if you're rewriting the code.

**Comment by baldurk** [comment]
From email discussion this extension is extremely badly designed to an extent that I had not fully appreciated, in particular that external textures are considered an entirely separate fundamental type of texture of the degree that they are accessed via a different type in shaders. As opposed to every other graphics API and several other supported GL extensions where they are normal textures, but created/initialised in a different way to perform the import operation.

This forces an implementation as this PR begins to do that must add support for this fake texture type throughout the GL backend and possibly further into the API abstraction (either with unknown real type or N different new fake types of texture to account for the actual types), as well as using external platform-specific APIs (which are likely unreliable on Android) for creating the textures on replay. The only other alternative that would be to somehow transform the captured commands and patch the use of external textures in shaders into normal textures, which can likely never be done robustly.

The conclusion I've come to is that this means a significant amount of implementation cost and maintenance burden which I do not feel is justified by the value of supporting this extension. While this is not a final written-in-stone decision that can never be reconsidered, I do not see a path forward where there would be an implementation of this extension that I would consider acceptable given the current situation. Since I also do not see that situation significantly changing, for now I consider this feature not something I ever plan to implement or support. Of course since RenderDoc is open source anyone is welcome to make whatever changes they wish locally, but this is not something I will accept upstream.

This is not a unique circumstance, there are a number of GL extensions which are not supported due to the cost of implementation and maintenance not being outweighed by the value of supporting them, and indeed there are a handful of things that are legal in core GL which are simply not supported and never will be due to them being extremely bad design decisions in GL and being far too costly or problematic to be worth supporting.

File: renderdoc/driver/gl/egl_dispatch_table.h
Comment (listevemeta) [inline renderdoc/driver/gl/egl_dispatch_table.h:73]: @lfedorov I am interested to have your change for our local usage for Android device. But this `PFNEGLGETNATIVECLIENTBUFFERANDROIDPROC` isn't defined in driver/gl/official
/eglext.h. Thus it doesn't compile. How can I get it compiled properly? Thanks.

**Pull request closed without merging**
